proposals/0127-get-sysvar-syscall.md

```diff
@@ -0,0 +1,198 @@
+entire data structure may still be too heavy to copy into a BPF context,
+fragments of the data can be made accessible via calls like
+`SlotHashes::get_slot(slot)` and `StakeHistory::get_entry(epoch)`.
+
+Additionally, as per [SIMD-0088](./0088-enable-core-bpf-programs.md), an effort
+is underway to reduce code complexity of the Solana runtime by porting native
+programs to Core BPF programs. These programs must be 100% backwards compatible,
+therefore requiring access to this sysvar data via syscalls and _not_ via
+providing the account directly. Requiring the account would break the programs'
+ABIs.
+
+### Reducing Syscall Bloat
+
+Currently, sysvar data layouts are very difficult to change. When called, the
+syscall handler will copy the entire sysvar to the given memory location.
+
+Notably, BPF programs do not inform the syscall handler of the expected size
+of the sysvar. BPF programs thus effectively assume that the requested sysvars
+will forever have the same size.
+
+This prevents future upgrades that shrink or extend sysvars. In the same
+spirit, if a new sysvar is created, a new syscall must be created.
+
+Furthermore, if certain queries are requred by BPF programs - such as
+`SlotHashes::get_slot()` and `StakeHistory::get_entry()` - those single-element
+retrievals will _also_ require new syscalls.
+
+It's worth noting that at the time of this writing, the work required for
+[SIMD-0088](./0088-enable-core-bpf-programs.md) will require at least five new
+syscalls if the existing design is not ratified.
+
+This tightly-coupled relationship between sysvars and syscalls will perpetually
+swell the syscall interface as changes are made, making maintainence and
+reimplementation difficult.
+
+## Alternatives Considered
+
+Other mechanisms considered that provide sysvar data to BPF programs are as
+follows:
+
+### Sysvar-Specific Syscalls
+
+This approach would involve changing nothing. We'd continue on the same course,
+directly coupling new sysvar fetch functions to new syscalls.
+
+### New VM Memory Regions
+
+A new memory region in the VM could be introduced to store sysvar data, which
+could then be made available to BPF programs through an interface around raw
+pointers, which loads the data similarly to the interface proposed herein.
+
+This approach would depend on direct mapping of host memory to VM memory,
+which has been in development for some time and requires further research
+and development to introduce safely to the VM ABI. This extensive R&D period
+would mean Core BPF program initiatives would remain blocked.
+
+Without direct mapping, sysvar data would be fully copied to this memory
+region each time a VM is created (currently one per transaction processed),
+causing a degradation in runtime performance and likely reducing compute
+by far less compared to the compute savings posed by this proposal.
+
+## New Terminology
+
+N/A.
+
+## Detailed Design
```

Comment:
Sorry, I guess I'm not entirely sure how much of our syscall interface is a specification or Agave-specific. For example, do all invoke contexts look the same? If so, we can expand this to say "anything from the invoke context is retrievable", but I was sort of under the impression that invoke context and sysvar cache can look different amongst clients. FD might serve syscalls from some other data source entirely when it comes to sysvars. Nobody says FD has to specifically have a sysvar _cache_. Right?

Reply:
Ok fair enough. Thanks for all of that context, it's helpful for this and other efforts!

I'll include the sysvars that are accessible by the new syscall in the proposal. As for your points in the earlier message:

> 1. What about the sysvars that are not in the sysvar cache? (Of which there are a bunch) (example: "the sysvar getter does not expose sysvars that are not in the sysvar cache)

I don't particularly see a valid reason for _not_ including data outside of the sysvar cache, but being specific about it. For example, I think it's reasonable to say things from the invoke context - instructions, recent blockhash, etc. - could be made available from this getter. Perhaps everything from the sysvar cache and select items from the invoke context?

> 2. What about any newly added sysvars? (example verbiage: "Unless specified otherwise, any sysvars that are introduced into the sysvar cache in the future also become visible via the sysvar getter")

Yeah, I think this should go in as-is.

> 3. The sysvar cache contains logical objects, not a byte serialization of these objects. Underspecifying this can lead to a consensus issue. When restoring the sysvar cache from the account DB, the sysvar cache ignores trailing data when deserializing. So does the sysvar getter include or exclude this trailing data?

I would lean towards no, it should not include it. It should include whatever's available to transaction processing, and if in the future we decide the **sysvar cache** _shouldn't_ ignore those trailing bytes, they're automatically made available to this syscall via the cache.

Reply:
imo we should specify a list of sysvars that the syscall supports and leave it at that

i was regarding the sysvar cache as an implementation detail that shouldnt be explicitly referenced in the specification of this at all, but it sounds like we need to make this explicit (and expect that firedancer will copy the sysvar cache mechanism exactly) because the cache may have consistency issues, so the source of the data could affect consensus?